cedar-policy-validator/src/typecheck.rs

```diff
@@ -441,24 +457,100 @@ impl<'a> Typechecker<'a> {
 
         // For every action compute the cross product of the principal and
         // resource applies_to sets.
-        all_actions.flat_map(|action| {
+        all_actions.flat_map(move |action| {
             action
                 .applies_to
                 .applicable_principal_types()
-                .flat_map(|principal| {
+                .flat_map(move |principal| {
                     action
                         .applies_to
                         .applicable_resource_types()
-                        .map(|resource| RequestEnv {
+                        .map(move |resource| RequestEnv {
                             principal,
                             action: &action.name,
                             resource,
                             context: &action.context,
+                            principal_slot: None,
+                            resource_slot: None,
                         })
                 })
         })
     }
 
+    /// Given a request environment and a template, return new environments
+    /// formed by instantiating template slots with possible entity types.
+    fn link_request_env<'b>(
+        &'b self,
+        env: RequestEnv<'b>,
+        t: &'b Template,
+    ) -> impl Iterator<Item = RequestEnv> + 'b {
+        self.possible_slot_instantiations(
+            t,
+            SlotId::principal(),
+            env.principal,
+            t.principal_constraint().as_inner(),
+        )
+        .flat_map(move |p_slot| {
+            self.possible_slot_instantiations(
+                t,
+                SlotId::resource(),
+                env.resource,
+                t.resource_constraint().as_inner(),
+            )
+            .map(move |r_slot| RequestEnv {
+                principal: env.principal,
+                action: env.action,
+                resource: env.resource,
+                context: env.context,
+                principal_slot: p_slot.clone(),
+                resource_slot: r_slot.clone(),
+            })
+        })
+    }
+
+    /// Get the entity types which could instantiate the slot given in this
+    /// template based on the policy scope constraints. We use this function to
+    /// avoid typechecking with slot bindings that will always be false based
+    /// only on the scope constraints.
+    fn possible_slot_instantiations(
```

Comment:
Would be more consistent to stick with the "link" terminology rather than switching to "instantiation" here.